healthx/handler_test.go

```diff
@@ -36,49 +36,133 @@ import (
-	assert.Equal(t, `{"errors":{"test":"not alive"}}`, strings.TrimSpace(string(out)))
-
-	alive = nil
-	response, err = c.Get(ts.URL + ReadyCheckPath)
-	require.NoError(t, err)
-	require.EqualValues(t, http.StatusOK, response.StatusCode)
-	require.NoError(t, json.NewDecoder(response.Body).Decode(&versionBody))
-	require.EqualValues(t, versionBody.Version, handler.VersionString)
+		}
+
+		router := httprouter.New()
+
+		handler.SetHealthRoutes(router, true)
+		handler.SetVersionRoutes(router)
+
+		ts := httptest.NewServer(router)
+
+		c := http.DefaultClient
+
+		var healthBody swaggerHealthStatus
+		response, err := c.Get(ts.URL + AliveCheckPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, json.NewDecoder(response.Body).Decode(&healthBody))
+		assert.EqualValues(t, "ok", healthBody.Status)
+
+		var versionBody swaggerVersion
+		response, err = c.Get(ts.URL + VersionPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, json.NewDecoder(response.Body).Decode(&versionBody))
+		require.EqualValues(t, versionBody.Version, handler.VersionString)
+
+		response, err = c.Get(ts.URL + ReadyCheckPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusServiceUnavailable, response.StatusCode)
+		out, err := ioutil.ReadAll(response.Body)
+		require.NoError(t, err)
+		assert.EqualValues(t, "ok", healthBody.Status)
+		assert.Equal(t, `{"errors":{"test":"not alive"}}`, strings.TrimSpace(string(out)))
+
+		alive = nil
+		response, err = c.Get(ts.URL + ReadyCheckPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, json.NewDecoder(response.Body).Decode(&versionBody))
+		require.EqualValues(t, versionBody.Version, handler.VersionString)
+
+	})
+
+	t.Run("case:with middleware", func(t *testing.T) {
+		// separate set of `handler` and `alive` as the `alive` would need to be reset after the last test in the above
+		alive := errors.New("not alive")
+		handler := &Handler{
+			H:             herodot.NewJSONWriter(nil),
+			VersionString: "test version",
+			ReadyChecks: map[string]ReadyChecker{
+				"test": func(r *http.Request) error {
+					return alive
+				},
+			},
+		}
+
+		// middleware to assert and test before the request is completed
+		testMiddleware := func(t *testing.T, assertFunc func(*testing.T, http.ResponseWriter, *http.Request)) func(next http.Handler) http.Handler {
+			return func(h http.Handler) http.Handler {
+				return http.HandlerFunc(func(writer http.ResponseWriter, req *http.Request) {
+					writer.Header().Add(mockHeaderKey, mockHeaderValue)
+					assertFunc(t, writer, req)
+					h.ServeHTTP(writer, req)
+				})
+			}
+		}
+
+		router := httprouter.New()
+
+		handler.SetHealthRoutes(router, true,
+			WithMiddleware(testMiddleware(t, func(t *testing.T, rw http.ResponseWriter, r *http.Request) {
+				assert.Equal(t, "GET", r.Method)
```

Comment:
Maybe we can add more assertions here, such as request path and request headers.